### DISCUSSION TITLE
Best interpretation of this BNF grammar?

### DISCUSSION THREAD

**Discussion by rowlesmr** [pyparsing/pyparsing#494] [Q&A]
I'm trying to implement [this piece of grammar](https://www.iucr.org/__data/assets/text_file/0009/112131/CIF2-ENBF.txt)

```
# { } : ZeroOrMore
# ( ) : a group
# | : MatchFirst
# - : exception: it matchs group on it's left, but not on its right.
# , : And

wsdelim-string = ( non-blank-char, { non-blank-char} ) - ( ( ( data-token | save-token ), { non-blank-char} ) | loop-token ) ;
```

So far, I've got
```python
ascii_lowercase = 'abcdefghijklmnopqrstuvwxyz'
ascii_uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
alphas = ascii_uppercase + ascii_lowercase
digits = '0123456789'
punctuation = r"""!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~"""
anp= alphas + digits+ punctuation 

data_token = CaselessLiteral("data_")
save_token = CaselessLiteral("save_")
loop_token = CaselessLiteral("loop_")

wsdelim_string = OneOrMore(anp) + ~(((data_token | save_token) + ZeroOrMore(anp)) | loop_token)

pass1 = "hello"
pass2 = "world"
pass3 = "data"
pass4 = "loop"
fail1 = "data_test"
fail2 = "data_"
fail3 = "save_frame"
fail4 = "loop_"
```

But I don't thing that's right.

Essentially, I want to match a group of one or more chars, unless they match something like `data_somechars`, or `save_`, in which case, I want the rule to fail.

**Comment by rowlesmr** [comment]
gah.

It should be
```python
wsdelim_string = ~(((data_token | save_token) + ZeroOrMore(anp)) | loop_token) + OneOrMore(anp)
```
shouldn't it....

**Comment by ptmcg** [comment]
Use `CaselessKeyword` instead of `CaselessLiteral`, and pyparsing will verify that you only match “data_” and not “data_1” or “data_set” or “data_that_i_downloaded_from_the_internet”.

 

I also scanned that document, and there is a lot of BNF around quoted strings. I think if you use the new pyparsing `python_quoted_string` builtin, it will cover all the bases (and you won’t need the single-quote, etc. pieces).

 

 

From: Matthew Rowles ***@***.***> 
Sent: Friday, July 7, 2023 10:39 AM
To: pyparsing/pyparsing ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [pyparsing/pyparsing] Best interpretation of this BNF grammar? (Discussion #494)

 

I'm trying to implement this piece of grammar <https://www.iucr.org/__data/assets/text_file/0009/112131/CIF2-ENBF.txt> 

# { } : ZeroOrMore
# ( ) : a group
# | : or
# - : exception: it matchs group on it's left, but not on its right.
# , : concatenation
 
wsdelim-string = ( alphas, { alphas } ) - ( ( ( data-token | save-token ), { alphas } ) | loop-token ) ;

So far, I've got

# alphas as per pyparsing
data_token = CaselessLiteral("data_")
save_token = CaselessLiteral("save_")
loop_token = CaselessLiteral("loop_")
 
wsdelim-string = OneOrMore(alphas) + ~(((data_token | save_token) + ZeroOrMore(alphas )) | loop_token)

But I don't thing that's right.

Essentially, I want to match a group of one or more chars, unless they match something like data_somechars, or save_, in which case, I want the rule to fail.

**Comment by rowlesmr** [comment]
> I also scanned that document, and there is a lot of BNF around quoted strings. I think if you use the new pyparsing `python_quoted_string` builtin, it will cover all the bases (and you won’t need the single-quote, etc. pieces).

I'm using your quote builtins for those. They are very useful.

I need `CaselessLiteral`, as when `data_token` appears when it should, it is of the form `data_token + OneOrMore(anp)`, something like "data_sometext`.
